tests/unit/s2n_client_hello_retry_test.c

```diff
@@ -0,0 +1,748 @@
+         * is called within this function */
+        EXPECT_SUCCESS(s2n_server_hello_retry_send(server_conn));
+
+        s2n_tls13_connection_keys(server_keys, server_conn);
+        uint8_t hash_digest_length = server_keys.size;
+
+        /* Obtain the recreated transcript hash */
+        struct s2n_hash_state server_hash, server_hash_state;
+        uint8_t server_digest_out[S2N_MAX_DIGEST_LEN];
+        GUARD(s2n_handshake_get_hash_state(server_conn, server_keys.hash_algorithm, &server_hash_state));
+
+        GUARD(s2n_hash_new(&server_hash));
+        GUARD(s2n_hash_copy(&server_hash, &server_hash_state));
+        GUARD(s2n_hash_digest(&server_hash, server_digest_out, hash_digest_length));
+        GUARD(s2n_hash_free(&server_hash));
+
+        struct s2n_blob server_blob;
+        EXPECT_SUCCESS(s2n_blob_init(&server_blob, server_digest_out, hash_digest_length));
+
+        EXPECT_SUCCESS(s2n_stuffer_wipe(&client_conn->handshake.io));
+        EXPECT_SUCCESS(s2n_stuffer_copy(&server_conn->handshake.io, &client_conn->handshake.io,
+                                        s2n_stuffer_data_available(&server_conn->handshake.io)));
+
+        EXPECT_SUCCESS(s2n_set_connection_hello_retry_flags(client_conn, 1));
+        /* Client receives the HelloRetryRequest mesage, s2n_server_hello_retry_recreate_transcript
+         * is called within this function */
+        EXPECT_SUCCESS(s2n_server_hello_recv(client_conn));
+
+        s2n_tls13_connection_keys(client_keys, client_conn);
+        hash_digest_length = client_keys.size;
+
+        /* Obtain the recreated transcript hash */
+        struct s2n_hash_state client_hash, client_hash_state;
+        uint8_t client_digest_out[S2N_MAX_DIGEST_LEN];
+        GUARD(s2n_handshake_get_hash_state(client_conn, client_keys.hash_algorithm, &client_hash_state));
+
+        GUARD(s2n_hash_new(&client_hash));
+        GUARD(s2n_hash_copy(&client_hash, &client_hash_state));
+        GUARD(s2n_hash_digest(&client_hash, client_digest_out, hash_digest_length));
+        GUARD(s2n_hash_free(&client_hash));
+
+        struct s2n_blob client_blob;
+        EXPECT_SUCCESS(s2n_blob_init(&client_blob, client_digest_out, hash_digest_length));
+
+        /* The transcript hash recreated MUST be the same */
+        S2N_BLOB_EXPECT_EQUAL(client_blob, server_blob);
+
+        EXPECT_SUCCESS(s2n_config_free(client_config));
+        EXPECT_SUCCESS(s2n_config_free(server_config));
+        EXPECT_SUCCESS(s2n_connection_free(server_conn));
+        EXPECT_SUCCESS(s2n_connection_free(client_conn));
+        EXPECT_SUCCESS(s2n_cert_chain_and_key_free(tls13_chain_and_key));
+    }
+
+    /* In this test, the client initiates a handshake with an X25519 share.
+     * The server, however does not support x25519 and prefers P-256.
+     * The server then sends a HelloRetryRequest that requires the
+     * client to generate a key share on the P-256 curve. */
+    if (s2n_is_evp_apis_supported()) {
```

Comment:
Are these "self-talk lite" tests still necessary? What do they provide that isn't covered by other unit tests + the self-talk test?

Reply:
No they are covered in the "self-talk" test, will get rid of it.